### PR TITLE
RoPE length extrapolation with interpolation

### PR DESCRIPTION
**Pull request by MaximumEntropy**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: NLP

# Changelog 
- Added the ability to do length extrapolation with RoPE via interpolation.
- Added a continue training script for GPT

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py [hunk 1]
```diff
@@ -21,13 +21,17 @@
 
 
 class RotaryEmbedding(nn.Module):
-    def __init__(self, dim):
+    def __init__(self, dim, seq_len_interpolation_factor=None):
```
Comment (ekmb) [inline nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py:24]: docstring?

File: examples/nlp/language_modeling/conf/megatron_gpt_config.yaml [hunk 2]
```diff
@@ -83,6 +83,7 @@ model:
   share_embeddings_and_output_weights: True # Share embedding and output layer weights.
   overlap_p2p_comm: False # Overlap p2p communication with computes. This argument is valid only when `virtual_pipeline_model_parallel_size` is larger than 1
   batch_p2p_comm: True # Batch consecutive inter-peer send/recv operations. This argument is valid only when `virtual_pipeline_model_parallel_size` is larger than 1
+  seq_len_interpolation_factor: null
```
Comment (ekmb) [inline examples/nlp/language_modeling/conf/megatron_gpt_config.yaml:86]: could you add a comment/paper ref here please?

File: examples/nlp/language_modeling/megatron_gpt_continue_training.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
```
Comment (ekmb) [inline examples/nlp/language_modeling/megatron_gpt_continue_training.py:1]: 2023

File: nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py (hunk 1 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py]: Done!

File: examples/nlp/language_modeling/conf/megatron_gpt_config.yaml (hunk 2 above)
Comment (MaximumEntropy) [reply examples/nlp/language_modeling/conf/megatron_gpt_config.yaml]: Done!

File: examples/nlp/language_modeling/megatron_gpt_continue_training.py (hunk 3 above)
Comment (MaximumEntropy) [reply examples/nlp/language_modeling/megatron_gpt_continue_training.py]: Done!

**Review by ekmb — APPROVED** [review]
Thank you!

File: examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml [hunk 4]
```diff
@@ -60,6 +60,8 @@ model:
   activations_checkpoint_num_layers: null # not used with 'selective'
   answer_only_loss: False # not used right now
   gradient_as_bucket_view: False
+  seq_len_interpolation_factor: null # if not None, seq_len_interpolation_factor will match the base model's value
```
Comment (yidong72) [inline examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml:63]: add some explanation about how interpolation factor translate to longer sequences.
e.g.  factor = 2,  sequence length x 2

File: examples/nlp/language_modeling/megatron_gpt_continue_training.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    NLPSaveRestoreConnector,
+    PipelineMixedPrecisionPlugin,
+)
+from nemo.core.config import hydra_runner
+from nemo.utils import AppState, logging
+from nemo.utils.exp_manager import exp_manager
+from nemo.utils.model_utils import inject_model_parallel_rank
+
+
+def _modify_config(gpt_cfg, cfg, add_cfg_to_tree=False):
```
Comment (yidong72) [inline examples/nlp/language_modeling/megatron_gpt_continue_training.py:38]: these `modify config`,  `load_from_nemo`, `load_from_checkpoint_dir`, `validate_checkpoint_loading_args` functions are the same as in the SFT code. can we put them into an utility file?

File: nemo/collections/nlp/modules/common/megatron/language_model.py [hunk 6]
```diff
@@ -559,7 +562,9 @@ def __init__(
             assert 0 < rotary_percentage <= 1
             if rotary_percentage < 1:
                 rotary_dim = int(rotary_dim * rotary_percentage)
-            self.rotary_pos_emb = RotaryEmbedding(rotary_dim)
+            self.rotary_pos_emb = RotaryEmbedding(
```
Comment (yidong72) [inline nemo/collections/nlp/modules/common/megatron/language_model.py:565]: maybe not in this PR. we need to add seq_len_interpolation_factor for all the models that uses RoPe.

File: examples/nlp/language_modeling/megatron_gpt_continue_training.py (hunk 5 above)
Comment (MaximumEntropy) [reply examples/nlp/language_modeling/megatron_gpt_continue_training.py:38]: They are almost the same not 100% identical. The issue is each one (SFT vs continued training) modifies some common attributes like `data`, `optim`, but also a few different things.

**Pull request merged**